docs/hugo/content/design/ADR-2023-01-Conversion-Augmentation.md

```diff
@@ -0,0 +1,135 @@
+    }
+}
+```
+
+**Pro**: Minimizes the amount of code we need to generate because we only need an extra clause in the existing conversion methods.
+
+**Con**: Requires the user to correctly provide the right type parameter when implementing the interface to augment the property conversions. For example, in this above situation, only an implementation of `genruntime.Assignable[*v20210101s.BatchAccount]` will be invoked; if the user instead implemented `genruntime.Assignable[**v20210101.BatchAccount]` the code would compile, but be silently ignored.
+
+**Con**: Go prohibits doing a type cast on a concrete type, so we need to stash `source` in an intermediate variable first.
+
+### Code-gen interface implementation
+
+Instead of using a generic interface, we can define a new interface for each version of each resource. This interface will be implemented by the user and will be called by the code generator.
+
+For example, in the `v1beta20210101` package, we'll generate an interface for the conversion hook:
+
+``` go
+type assignableBatchAccount interface {
+	AssignPropertiesToBatchAccount(dst *v20210101s.BatchAccount) error
+	AssignPropertiesFromBatchAccount(src *v20210101s.BatchAccount) error
+}
+```
+
+This interface would then be detected and used if present:
+
+``` go
+var sourceAsAny any = source
+if src, ok := sourceAsAny.(assignableBatchAccount); ok {
+    err := src.AssignPropertiesFromBatchAccount(source)
+    if err != nil {
+        return errors.Wrap(
+            err, 
+            "calling custom AssignFrom() for conversion from v20210101s.BatchAccount")
+    }
+}
+```
+
+**Pro**: The interface is only accessible (visible) within the package and has the correct parameter types included. It's not possible for the user to *think* they've implemented the hook point and have it silently ignored. Either the compiler will complain, or the code will work.
+
+**Con**: We don't have any other interfaces being defined by the code generator, so this is a new pattern. (We do have interface implementations).
+
+**Con**: Go prohibits doing a type cast on a concrete type, so we need to stash `source` in an intermediate variable first.
+
+### Configuration implementation
+
+Add new configuration where we can specify that a particular resource needs tweaking. For the Batch account, above, this might look like:
+
+``` yaml
+objectModelConfiguration:
+  batch:
+    2021-01-01:
+      BatchAccount:
+        $customConversion: true
+```
+
+When found, we'd directly generate a call to the conversion functions were needed:
+
+``` go
+err := source.AssignPropertiesFromBatchAccount(source)
+if err != nil {
+    return errors.Wrap(
+        err, 
+        "calling custom AssignFrom() for conversionfrom v20210101s.BatchAccount")
+}
+```
+
+**Pro**: The configuration is very explicit and easy to understand.
+
+**Con**: We need to add a new configuration option and the code to process it.
```

Comment:
Where would the implementation of this function live? It's authored by the user so the generated code won't compile on its own? That's sorta a downside as AFAIK up until now an invariant of the code generator has been that it produces code that compiles.

Reply:
Agreed. I've called this out explicitly. The implementation would be written by the end user, and until that was done nothing would work.